src/random.rs

```diff
@@ -0,0 +1,174 @@
+use crate::context::Context;
+use rand::SeedableRng;
+use std::any::{Any, TypeId};
+use std::cell::{RefCell, RefMut};
+use std::collections::HashMap;
+
+/// Use this to define a unique type which will be used as a key to retrieve
+/// an independent rng instance when calling `.get_rng`.
+#[macro_export]
+macro_rules! define_rng {
+    ($random_id:ident) => {
+        struct $random_id {}
+
+        impl $crate::random::RngId for $random_id {
+            // TODO: This is hardcoded to StdRng; we should replace this
+            type RngType = rand::rngs::StdRng;
+
+            fn get_name() -> &'static str {
+                stringify!($random_id)
```

Comment:
Note that there are several ways in which one could end up having duplicate names. I wonder if we should store the type_id in the box and assert at runtime if someone tries to do `get_rng<T>` when `get_rng<U>` has already been called and T.get_name() == U.get_name()

Reply:
Alright, I added this to the macro which should ensure this actually fails at compile time:

```rust
pub static $random_id: () = ();
```

Reply:
The downside to the static variable solution is potential collisions with other types (which we could reduce by mangling the name a bit), and that it only works if people use the macro (which of course they should). Maybe the runtime solution is better as we would only check when `get_rng` is called below and when the generator needs to be created (basically on first sample after seeding/reseeing)?

Reply:
I mean, do you actually want two types across to have the same name? It's true that runtime check is safer without the macro though, that's probably enough of a reason to prefer it

Reply:
On second thought after implementing the runtime check, it is quite a bit more code / requires storing the name etc. and I lean towards leaving it as is – 

I pushed a version that namespaces the type name a bit (I couldn't figure out how to do it without importing `paste`)